lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -2339,48 +2374,76 @@ void addIntelFPGADecorations(SPIRVEntry *E, DecorationsInfoVec &Decorations) {
     case DecorationUserSemantic:
       E->addDecorate(new SPIRVDecorateUserSemanticAttr(E, I.second));
       break;
-    case DecorationMemoryINTEL:
-      E->addDecorate(new SPIRVDecorateMemoryINTELAttr(E, I.second));
-      break;
+    case DecorationMemoryINTEL: {
+      if (M->isAllowedToUseExtension(
+              ExtensionID::SPV_INTEL_fpga_memory_attributes))
```

Comment:
Before your change this function early exited if `SPV_INTEL_fpga_memory_attributes` and `SPV_INTEL_fpga_memory_accesses` both not allowed. Now there is checks for only one extension for most of decorations. Looks like it changes the previous logic, was it an intention?

Reply:
Yes, it was an intention, there are some decorations that are allowed only for `SPV_INTEL_fpga_memory_attributes` and some for `SPV_INTEL_fpga_memory_accesses`, so previously implementation wasn't separating them, which is a bug. From the description:
```
Also this patch brings a bit of refactoring of add Intel FPGA decoration functions,
since their infrustructure can (and should) be reused here, solving an issue of
incorrect allowed FPGA memory SPIR-V extensions processing.
```
Sorry for a mixing this up in a single patch, I was just trying to fit this patch in the existing code and 'handle FPGA memory decorations' infrastructure is looking natural to be reused. And since I was going to reuse this infrustructure I fixed this bug as well.